primitives/payment-handler/src/fee_handler.rs

```diff
@@ -46,118 +48,137 @@ impl<AccountId, Balance> Payment<AccountId, Balance> {
 		payer: &Self::AccountId,
-		payment_asset: Self::AssetId,
+		payment: Self::PaymentKind,
 		base_fee: Self::Balance,
 		tournament_id: &Self::TournamentFeeIdentifier,
 		affiliate_id: &Self::AffiliateFeeIdentifier,
 		treasury_pot: &Self::AccountId,
 	) -> Result<(), DispatchError> {
 		// The credit may be in any asset as implemented by `WithdrawAsset`.
-		let fee_credit = W::withdraw_credit(payer, payment_asset, base_fee)?;
-
-		let remaining_credit =
-			Self::try_propagate_tournament_fee(fee_credit, payer, tournament_id)?;
+		if let Some(fee_credit) = W::withdraw_credit(payer, payment.clone(), base_fee)? {
+			let remaining_credit =
+				Self::try_propagate_tournament_fee(fee_credit, payer, tournament_id)?;
 
-		let remaining_credit2 =
-			Self::try_propagate_chain_fee(remaining_credit, payer, affiliate_id)?;
+			let remaining_credit2 =
+				Self::try_propagate_chain_fee(remaining_credit, payer, affiliate_id)?;
 
-		Self::deposit_into_treasury(treasury_pot, remaining_credit2)
+			Self::deposit_into_treasury(treasury_pot, remaining_credit2)
+		} else {
+			Ok(())
+		}
 	}
 
 	fn withdraw_and_deposit_into_treasury(
 		who: &Self::AccountId,
-		asset_id: Self::AssetId,
+		payment: Self::PaymentKind,
 		treasury_pot: &Self::AccountId,
 		amount: Self::Balance,
 	) -> Result<(), DispatchError> {
-		let credit = W::withdraw_credit(who, asset_id, amount)?;
-		Self::deposit_into_treasury(treasury_pot, credit)
+		if let Some(credit) = W::withdraw_credit(who, payment, amount)? {
+			Self::deposit_into_treasury(treasury_pot, credit)
+		} else {
+			Ok(())
+		}
 	}
 }
 
-impl<AccountId, Assets, W, Affiliate, Tournament>
-	AssetGameFeeHandler<AccountId, Assets, W, Affiliate, Tournament>
+impl<AccountId, Assets, W, Affiliate, MaxAffiliates, Tournament>
+	AssetGameFeeHandler<AccountId, Assets, W, Affiliate, MaxAffiliates, Tournament>
 where
-	Assets: fungibles::Inspect<AccountId, Balance = W::Balance, AssetId = W::AssetId>
-		+ fungibles::Balanced<AccountId>,
-
+	Assets: fungibles::Balanced<AccountId, Balance = W::Balance>,
 	W: WithdrawCredit<
 		AccountId = AccountId,
 		Assets = Assets,
 		Credit = fungibles::Credit<AccountId, Assets>,
 	>,
 
-	Affiliate: DistributeFee<AccountId = W::AccountId, Balance = W::Balance>,
-	Tournament:
-		DistributeFee<AccountId = AccountId, Balance = W::Balance, MaxDistributions = ConstU32<1>>,
+	Affiliate: DistributeFee<
+		AccountId = AccountId,
+		Balance = W::Balance,
+		FeeDistribution = AffiliateFeeDistribution<AccountId, W::Balance, MaxAffiliates>,
+	>,
+	Tournament: DistributeFee<
+		AccountId = AccountId,
+		Balance = W::Balance,
+		FeeDistribution = TournamentFeeDistribution<AccountId, W::Balance>,
+	>,
 {
 	/// Distributes an already withdrawn `fee_credit` to the affiliates of `account`.
 	///
 	/// Returns the remaining `fee_credit` after this operation.
 	fn try_propagate_chain_fee(
-		fee_credit: W::Credit,
-		account: &W::AccountId,
+		fee_credit: fungibles::Credit<AccountId, Assets>,
```

Comment:
Can we use `W::Credit` again? - it is easier to read. I think you have changed it because you had to in the first try.

There are a few other cases in this file, where this applies too.